### ISSUE TITLE
match incompatible for TypeScript 3.3.4 later

### ISSUE THREAD

**Issue by septs** [aikoven/typescript-fsa#73]
error info:
```
(property) AsyncActionCreators<string, void, Error>.done: {
    (payload: Success<string, void>, meta?: {
        [key: string]: any;
    } | null | undefined): Action<Success<string, void>>;
    type: string;
    match: (action: AnyAction) => action is Action<Success<string, void>>;
}

Argument of type '{ (payload: Success<string, void>, meta?: { [key: string]: any; } | null | undefined): Action<Success<string, void>>; type: string; match: (action: AnyAction) => action is Action<Success<string, void>>; }' is not assignable to parameter of type '{ (payload: { params: string; } & { result: void; }, meta?: { [key: string]: any; } | null | undefined): Action<{ params: string; } & { result: void; }>; type: string; match: (action: AnyAction) => action is Action<{ ...; } & { ...; }>; }'.
  Types of property 'match' are incompatible.
    Type '(action: AnyAction) => action is Action<Success<string, void>>' is not assignable to type '(action: AnyAction) => action is Action<{ params: string; } & { result: void; }>'.
      Type predicate 'action is Action<Success<string, void>>' is not assignable to 'action is Action<{ params: string; } & { result: void; }>'.
        Type 'Action<Success<string, void>>' is not assignable to type 'Action<{ params: string; } & { result: void; }>'.
          Type 'Success<string, void>' is not assignable to type '{ params: string; } & { result: void; }'.
            Type '{ params: string; } & { result?: void | undefined; }' is not assignable to type '{ params: string; } & { result: void; }'.
              Type '{ params: string; } & { result?: void | undefined; }' is not assignable to type '{ result: void; }'.
                Property 'result' is optional in type '{ params: string; } & { result?: void | undefined; }' but required in type '{ result: void; }'.

ts(2345)
```

errata code:

https://github.com/aikoven/typescript-fsa/blob/dc494592263b31269094fdcc1a80f83c7616ec8b/src/index.ts#L84-L90

if params is void returns never?

if result is void returns never?

```diff
diff --git a/src/index.ts b/src/index.ts
index 5e99796..e1dc9dc 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -82,11 +82,11 @@ export type ActionCreator<Payload> = {
 );

 export type Success<Params, Result> =
-  ({params: Params} | (Params extends void ? {params?: Params} : never)) &
-  ({result: Result} | (Result extends void ? {result?: Result} : never));
+  ({params: Params} | (Params extends void ? never : {params?: Params})) &
+  ({result: Result} | (Result extends void ? never : {result?: Result}));

 export type Failure<Params, Error> =
-  ({params: Params} | (Params extends void ? {params?: Params} : never)) &
+  ({params: Params} | (Params extends void ? never : {params?: Params})) &
   {error: Error};

 export interface AsyncActionCreators<Params, Result, Error = {}> {
```

**Comment by aikoven** [comment]
Thanks for the report. 

You may use version `2.5.0` until this is fixed.

**Comment by septs** [comment]
@aikoven not fixed to 3.0.0 version?

**Comment by aikoven** [comment]
3.0.0 is not stable yet. It has some experimental features that don’t yet work correctly in all setups.
